### DISCUSSION TITLE
SOTA autocomplete with dropdown

### DISCUSSION THREAD

**Discussion by AndreasK79** [magicbug/Cloudlog#808] [Show and tell]
I started looking for something better than a dropdown when it comes to loading a lot of selections in a form.

I ended up with trying out [selectize](https://github.com/selectize/selectize.js). You can specify a text-input, which then loads a result via ajax and presents it as a dropdown. like this:
![image](https://user-images.githubusercontent.com/6977712/104125008-63b38600-5354-11eb-932e-153fa9c46508.png)

In my test, I limited the result to 30 to get it snappy enough to use, as too many entries would slow down everything.

@magicbug The question now is which solution I should use. Should I use a plain txt-file to preg_grep from, or should I create a new sota-table in the database? I have both solutions working on my machine.  The SOTA list contains 152k entries. I don't know how often these get updated, so I don't know how much job it is to maintain a list.

I think something similar could be used for US Counties, as loading several thousand in a dropdown select would slow things down.

**Comment by magicbug** [comment]
Good question, I think we could just keep it a text file similar to the scp file that seems quite snappy and one less database call.

Counties need pretty much the same either text, json that can be checked against the "state" then pull up the available counties.

There is a DOK file @phl0 did point me at it but I have lost it!

**Comment by phl0** [comment]
You want to pull DOK data in from a list of valid entries?

**Reply by magicbug** [reply]
*nods*

**Reply by AndreasK79** [reply]
@phl0 Added DOK to the PR. Care to test?

**Reply by phl0** [reply]
Jepp. Will do and report.

**Reply by phl0** [reply]
It does look very good. Test completed successfully. Would like to see that feature in the upstream code.
How is update of SOTA/DOK info handled? Maybe we should add a cron job for that?
I did a minor fix against @AndreasK79's code: https://github.com/AndreasK79/Cloudlog/pull/2

**Reply by AndreasK79** [reply]
Thanks for testing @phl0. PR merged. Not sure how to do the update. Someone needs to update and host the files. They are not directly hosted now, as they need to be edited from the files available for download.

**Reply by phl0** [reply]
Did you pull the raw DOK data from my server? 
I see two possibilities:
- Either I provide the data in correct format by extending my scripts or
- The cron job/script pulls the data and formats it accordingly

**Reply by AndreasK79** [reply]
I used your csv file with only the first column. If you can provide a txt file in that format, autoupdate is possible.

**Reply by phl0** [reply]
Let me check. QRX

**Reply by phl0** [reply]
Script updated. See: https://www.df2et.de/cqrlog/dok.txt

**Comment by AndreasK79** [comment]
File it is then. I just need to tweak the css, as it has it's own file, and it would be nice to get it to match the rest of the theme. I'm nearly there for the Superhero theme:
![image](https://user-images.githubusercontent.com/6977712/104196504-24ed0100-5424-11eb-9092-429e538ea210.png)

@magicbug I guess Counties can be done the same way, but SOTA first. And I would need a list of Counties connected to states.

@phl0 if you can get me a text-file with the DOK's, the same can be done there.

**Reply by magicbug** [reply]
@AndreasK79 Inside the LoTW configuration file theres an XML list with them in the config.tq6 which is just a compressed XML file it actually has the list not only for US but also Alaska, Russia, Hawaii etc too

**Comment by phl0** [comment]
Have a look here:

https://www.df2et.de/cqrlog/

I did write some perl magic to pull current DOK data from the DARC website and parse it along with the clear text names etc.

**Comment by AndreasK79** [comment]
#831 added. Please test it out. It's just for SOTA now, but when it is done and merged, I'll look into counties and DOK's.
The CSS was matched as close as I managed. There are a few things I never could figure out. Any CSS experts around?

**Reply by AndreasK79** [reply]
The problem I'm having, is that I can't get the bright border when the input is selected. I've tried to study the CSS, and I think I found the correct setting, but it just doesn't work.
![Screenshot 2021-01-16 135503](https://user-images.githubusercontent.com/6977712/104812502-ef209180-5802-11eb-8f78-0c28ed16b7c7.png)
![Screenshot 2021-01-16 135451](https://user-images.githubusercontent.com/6977712/104812503-f051be80-5802-11eb-9765-6b22c713bc7e.png)

This happens on Blue, Superhero and Darkly.

**Comment by AndreasK79** [comment]
Now that the PR is merged, I've added another PR for this to be active in the edit dialog.

@phl0 I'll see if I can make a cron job for autoupdating DOK from your file. Thanks. Perhaps you could do the same for the SOTA as well? The full file is here: https://www.sotadata.org.uk/summitslist.csv All I need, is the first column in a text file.

**Reply by phl0** [reply]
Roger that. I wouldn't be a problem. But I guess it woukd be easier to integrate the conversion into the script executed by cron. I am happy do deliver bash/shell code for that.

**Reply by AndreasK79** [reply]
@phl0 That is also a possibility. 

@magicbug What's your opinion on this?

**Reply by phl0** [reply]
I guess something like 

```
wget -O - "https://www.sotadata.org.uk/summitslist.csv" | awk -F "," '{ print $1; }' > sota.txt
```

should do the job quite fine.

**Reply by AndreasK79** [reply]
It's been a while since I played with AWK. Those oneliners are nice :)

**Reply by AndreasK79** [reply]
@phl0 I pushed a PR now for autoupdate. I did it in PHP, but I guess your oneliner could have been ran in CRON just as easy.

**Reply by phl0** [reply]
Fine with me. Need to enable those cron jobs on my instance(s) one day... 😇✌️

**Comment by ryandzink** [comment]
@AndreasK79 this may just be my personal logging style but do you think there's a possibility here to, once the summit identifier is selected, pull the summit name, gridsquare, etc from the SOTA API and automatically insert them into the relevant fields as well?

**Reply by AndreasK79** [reply]
@ryandzink I'll make no promises, but I'll make a note and take a look. 

- I think this must be an option to turn on/off
- Where would you like summit name? Location?
- And grid should be placed in locator I would assume. (this needs to be calculated, as I don't see the grid itself in the data).

**Reply by ryandzink** [reply]
@AndreasK79 yes I agree it would be optional. I'm not even sure that anyone else feels like this would be useful 🤷‍♂️ 

Yes, I was thinking summit name in the Location field and 6 digit grid in Locator. I didn't think either one needed to be calculated, can't they be pulled directly from the API? e.g. https://api2.sota.org.uk/api/summits/W7A/AW-060 contains both the summit name and locator.

**Reply by AndreasK79** [reply]
@ryandzink you are right. I was looking, but my mind was caught up in gridref1 and gridref2. I see now that locator was specified further down. I'll see if I can figure out something when I find some time.

**Reply by AndreasK79** [reply]
@ryandzink #897 was just added. This can be turned on/off in user settings.

**Comment by hb9hnt** [comment]
Sorry, a bit late to the party but is there a plan to implement this for the creation of station locations, too? As a SOTA activator it would be awesome to have this as you create a new station for every activation and since station locations don't save without a grid locator you have to look up the locator every time (or use the QSO form as a workaround ;) )

Thanks and 73 de Beni HB9HNT
